### ISSUE TITLE
No support for PEM key import

### ISSUE THREAD

**Issue by vingtheory** [edjCase/ICP.NET#59]
So I have a Pxy Staking canister which got several admin methods which only works with my default identity installed in my ubuntu machine in the location '\\wsl.localhost\Ubuntu-20.04\home\ving\.config\dfx\identity\default'. 

As you may know, the file 'identity.pem' has the private key. 

Everyday, I run pxy's stats method (ex: runStats) in Ubuntu CLI and it runs in the default identity context of the above identity since that is the primary controller of the canister. It works just fine.

Ex: dfx canister --network ic call staking runStats

I am trying to automate this by using your ICP.Net library. I am using my default identity's Principal as "public key" and copied the base64 encoded string as the "private key". Then I am instantiating HttpAgent by passing the identity object. But for some reason, its throwing exception. 

Please follow the code snippet below. 

```
static async Task AdminUnstake()
{
    try
    {
        // public key
        // obtained from your identity's principal id by running following command
        // dfx identity get-principal
        // ex: ax3nc-yx45z-w6iew-d2hds-2gmuc-z4hkb-okg7h-4z6y5-3ach6-eclvr-yqe
        var principal = Principal.FromText("my-default-identity-principal");
        var hex = principal.ToHex();
        byte[] publicKeyBytes = ByteUtil.FromHexString(hex);
        var publicKey = DerEncodedPublicKey.FromEd25519(publicKeyBytes);

        // private key
        // copied from identity.pem file
        byte[] privateKeyBytes = Convert.FromBase64String("base-64-encoded-string-found-inside-identity.pem-file");
        
        var identity = new Ed25519Identity(publicKey, privateKeyBytes);

        var agent = new HttpAgent(identity);
        Principal canisterId = Principal.FromText("5xaxa-xaaaa-aaaal-abcpa-cai");
        var pxyStakingClient = new PxyStakingApiClient(agent, canisterId);
        var result = await pxyStakingClient.AdminUnstake(2829);
    }
    catch (Exception ex)
    {
        Console.WriteLine(ex.Message);
    }
}
```

The exception is not super informative. So, it didnt help much. 

However, I am posting it below for your reference. 

![image](https://user-images.githubusercontent.com/112404892/230756170-8dc3c181-31d0-4029-816f-3e7a9a8286b5.png)

Very impressive work on how the ClientGenerator and it just saves a LOT of time. Thank you!! :-)

**Comment by Gekctek** [comment]
The main issue that I see is that the Principal is NOT the public key. The principal is derived from the key, essentially a hash of it. 

Im working on something to where you can import your pem/private key and be able to get the public key from it

**Comment by vingtheory** [comment]
> 
Thanks for the response. 

Ah. I see. Somehow, I don't know how to obtain the public key from the private key and I thought Principal would help me get it. Probably, this is the part of my problem there, I guess. 

Also, please validate my code above for the private key and let me know if the logic is alright. 

Thank you.

**Comment by Gekctek** [comment]
@vingtheory Ive been stuck on this issue for a while. It breaks down to a couple things:
1) The dfx PEM file seems to create a Ecdsa key vs a ed22519. That is fine, but I am working on a new identity type for that
2) I have some code but the cryptography is above my head and I cant figure out why its not working quite yet

Im not sure if there is a workaround that I can think of so you might be stuck till I figure this out. I might reach out to some dfinity people

**Comment by vingtheory** [comment]
> @vingtheory Ive been stuck on this issue for a while. It breaks down to a couple things:
> 
> 1. The dfx PEM file seems to create a Ecdsa key vs a ed22519. That is fine, but I am working on a new identity type for that
> 2. I have some code but the cryptography is above my head and I cant figure out why its not working quite yet
> 
> Im not sure if there is a workaround that I can think of so you might be stuck till I figure this out. I might reach out to some dfinity people

Hey Sir. Its totally fine. I will also spend some time tinkering with the encryptions and visiting dev forums in coming days. Will let you know if anything comes up. Thank you.

**Comment by Gekctek** [comment]
Just wanted to give an update. Not much progress, but still working on it

**Comment by vingtheory** [comment]
Thank you for the prompt update. Appreciate it! There is a bounty from Dfinity to develop a file-share module (similar to a OneDrive desktop service which auto syncs the watched folders in the file system). I and @lightninglad working on this. We are currently covering the web side only. I am thinking your library will come handy when we get to work on desktop windows service. Probably we can collab on this. You got such a wonderful library which can be rewarding. Let me know. :-)

**Comment by Gekctek** [comment]
Sounds like fun. Ill try to help if you guys need it. 
This library benefits a lot from getting feedback from different projects so this is a unique use case, most have been just gaming so far

**Comment by vingtheory** [comment]
Thank you so much for your positive attitude to help in this regard. 😊

Also, I just received a notification that you have a commit to support PEM integration. Looks like you already got it. Cant wait to use it. Thanks in advance. 🙏

**Comment by Gekctek** [comment]
@vingtheory Alright. I had a breakthrough with these PEM files/key encodings
Try out version `3.0.0-beta.1`
I havent documented it quite yet or put in all the tests but wanted to see if this works for you before I polish it up

Youll want to try using the `IdentityUtil.FromPemFile` method to get it out of the PEM file
If you just have the private key I have added the `IdentityUtil.FromSecp256k1PrivateKey` or `IdentityUtil.FromEd25519PrivateKey` to derive the public keys if you dont have them

Let me know what you think, then ill make an non beta release and document it

**Comment by vingtheory** [comment]
Thank you for promptly getting this out. It will many of us. I will get to this next week and play around with what I was doing earlier. I will add some test cases and may issue PR. Can't wait to get to this soon.

**Comment by Gekctek** [comment]
Fixed by #64 
Updated docs to give examples of PEM import in the Agent README
@vingtheory let me know if there are issues or anything missing
[Version 3.0.0](https://github.com/edjCase/ICP.NET/releases/tag/3.0.0)

**Comment by vingtheory** [comment]
thank you so much for making this available. will create test cases soon and send your way!
